doc/man/gnunetd.1
.TH GNUNETD "1" "11 Nov 2006" "GNUnet"
.SH NAME
gnunetd \- The GNUnet daemon
.SH SYNOPSIS
.B gnunetd
[\fIOPTIONS\fR]
.SH DESCRIPTION
.PP
The GNUnet daemon, required for any GNUnet operations (searches, downloads, etc). gnunetd does not require root rights.
.TP
\fB\-c \fIFILENAME\fR, \fB\-\-config=FILENAME\fR
load config file (default: /etc/gnunet.conf)
.TP
\fB\-d\fR, \fB\-\-debug\fR
do not detach from the console (for debugging); log messages are written to stderr in addition to the logfile.
.TP
\fB\-h\fR, \fB\-\-help\fR
print help page
.TP
\fB\-l\fR, \fB\-\-loud\fR
print all log messages (to stderr).  Only works together with the \-d option.  Also disables logging to files.
.TP
\fB\-L \fILOGLEVEL\fR, \fB\-\-loglevel=LOGLEVEL\fR
change the loglevel. Possible values for LOGLEVEL are NOTHING, FATAL, ERROR, WARNING, INFO, STATUS and DEBUG.  Note that options in the configuration file take precedence over this option (the argument will be ignored in that case).
.TP
\fB\-u \fIUSER\fR, \fB\-\-user=USER\fR
run as user USER (and if available as group USER). Note that to use this option, you will probably have to start gnunetd as
root. It is typically better to directly start gnunetd as that user instead.
.TP
\fB\-v\fR, \fB\-\-version\fR
print the version number
.SH NOTES
Before you can share, search or download files from GNUnet you must start the GNUnet server, gnunetd.
If you start gnunetd as root (which is not required), you can use the \-u option such that GNUnet runs as a different user. If a group of the same name exists, GNUnet will also change to that group. If GNUnet cannot change its UID to USER or GID to USER, a warning will be printed and gnunetd will *continue* with the rights of the user that invoked gnunetd.
.SH FILES
.TP
/etc/gnunet.conf
gnunetd configuration file (default location).
.TP
/var/lib/GNUnet/.hostkey
Nodes GNUnet RSA private key.  Keep secret.
.TP
/var/lib/GNUnet/data/hosts/
Hostkeys.  GNUnet stores contact information and public keys of other nodes here. You may want to make this directory publically available.
.TP
/var/lib/GNUnet/data/fs/
Shared content (and indexing information) is stored in this directory.
.TP
/var/lib/GNUnet/data/credit/
Trust directory.  GNUnet stores economic information about other nodes here, in particular how much useful data we have received from which remote node.
.SH "REPORTING BUGS"
Report bugs by using mantis <https://gnunet.org/bugs/> or by sending electronic mail to <gnunet\-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunet\-gtk\fP(1), \fBgnunet\-insert\fP(1), \fBgnunet\-search\fP(1), \fBgnunet\-download\fP(1), \fBgnunetd.conf\fP(5), \fBgnunetd\fP(1)


ChangeLog
Mon Nov 16 09:25:47 CET 2009
	Allow passing of command-line arguments using GNUNET_ARGS
	environment variable (see Mantis #1467).

Mon Apr 27 22:18:19 CEST 2009
	Monitor low-traffic connections over a longer period
	of time before terminating them.

Fri Feb 27 22:05:50 MST 2009
	Releasing 0.8.0c.

Sun Feb  1 19:51:40 MST 2009
	Fixed Mantis #1429 (struct padding could cause sizeof
	to be different on AMD64, using gcc-ism "packed" to
	avoid this problem).

Sun Feb  1 00:37:16 MST 2009
	Fixed problem in DHT client protocol that could cause
	PUT requests from clients to be (frequently) lost.

Sun Nov  2 16:06:47 MST 2008
	Fixed problem with MySQL disconnects in one module
	impacting MySQL correctness in another module (caused
	crashes).

Sat Aug 16 00:18:34 CEST 2008
	Improved keyword extraction
	Fixed expiration time of keywords
	Releasing 0.8.0b.

Sun Jul 20 15:27:11 MDT 2008
	Releasing 0.8.0a.

Thu Jul 10 00:59:31 MDT 2008
	Fixed bug where datastore resizing would not 
	result in bloomfilter update during gnunet-update.

Sat Jun 28 23:43:00 MDT 2008
	Fixed bug preventing namespace advertisements from
	being processed.

Sat Jun 14 00:54:39 MDT 2008
	Releasing 0.8.0.

Fri Jun 13 23:51:00 MDT 2008
	GNUnet now compiles on OpenBSD.

Wed Jun 11 22:17:09 MDT 2008
	Made state of gnunet-auto-share persistent. 
	Added support for multiple directories.  Improved
	FSUI/ECRS APIs to reduce the number of threads needed
	for probing.  Increased number of concurrent probes
	allowed.

Sat Jun  7 17:54:49 MDT 2008
	Simplified namespaces a lot.  Shorter timeouts for
	trying to connect to gnunetd to prevent interactive
	apps from blocking too long when gnunetd is not
	running.

Sun Jun  1 21:11:54 MDT 2008
	Releasing 0.8.0pre1.

Sun Jun  1 13:41:27 MDT 2008
	Split of libgnunetpseudonym from libgnunetnamespace
	(with major API extension and changes).

Tue May  6 04:10:02 MDT 2008
	Added simple sanity check for [MODULES] values.
	Added "-s" and "-q" options to gnunet-peer-info.

Thu Apr 24 00:05:17 MDT 2008
	Simplified FSUI stop/abort/pause/resume API calls
	(no need to pass FSUI context anymore).

Mon Apr 21 21:05:20 MDT 2008
	Adding buffered IO for FSUI's serialize and
	deserialize code.

Sun Apr 20 20:01:20 MDT 2008
	Fixed bug in HTTP transport causing lonely messages
	(responses to GET) not to be delivered.

Mon Mar 24 21:15:36 MDT 2008
	Fixed various problems with downloading locally
	indexed large files (downloading large files from
	other peers should not have been a problem).
	
Sun Mar  2 15:33:41 MST 2008
	Added option to allow user to disable IPv6 support.
	Releasing 0.8.0pre0.

Sat Mar  1 21:14:02 MST 2008
	Fixed some dstore performance issues.
	Made transports alternate between trying IPv4 and IPv6
	if both are available (since one of them maybe
	misconfigured).

Mon Feb 25 23:25:48 MST 2008
	Cleanup of the DNS code: removed GNUNET_IPvXAddress
	and GNUNET_PRIP; centralized all DNS lookup code in 
	util/network/dns.c; improved IPv6 support for 
        network-client code.

Mon Feb 25 00:01:27 MST 2008
	Added asynchronous search and download methods for 
	ECRS library.  FSUI now can do with only one thread
	per search or download (until now, we had two 
	threads per search / download).

Tue Feb 19 20:35:28 MST 2008
	Updated database schemata to support O(1) operations
	even if there are N files under the same keywords.
	This breaks database backwards compatibility (and
	no migration code is provided).

Mon Feb 18 19:47:37 MST 2008
	Added new options -u and -s to gnunet-insert.

Sat Feb 16 21:37:33 MST 2008
	Implemented gnunet-auto-share for automatic
	sharing of directories.

Thu Feb 14 01:02:32 MST 2008
	Made HTTP transport work over IPv6 as well.

Wed Feb 13 22:59:07 MST 2008
	Modified mySQL and SQLite datastores to return diverse
	result sets during (partial) hash-based get iterations
	(should result in more diverse keyword result sets).

Tue Feb 12 23:54:34 MST 2008
	First shot at integrated hostlist HTTP server based
	on libmicrohttpd.

Sat Feb  9 22:06:23 MST 2008
	Minor changes to the ECRS/FSUI APIs for searches
	(no more timeouts and/or max-results).  Clients who
	need these features could easily implement them 
	directly (and they are not really useful to begin
	with).  
	Also, the IPv4 and IPv6 protocols are being integrated
	so that we only listen to one port (running both IPv4
	and IPv6 over the same port).  This break transport
	compatibility, but given that we're breaking protocol
	compatibility for file-sharing already, this may 
	actually be a bit of a good thing (since it will now
	be possible to tell if peers running the new GAP code
	are connected to other peers running the new GAP code).
	
Sun Feb  3 13:17:09 MST 2008
	Dramatic changes to the GAP implementation (breaking
	protocol compatibility).  Essentially, we can save
	a few bytes in each reply.  More importantly, the new
	code allows the searching client to specify a set of
	replies that are not desired (hopefully helping to
	dramatically increase the diversity of search replies
	obtained over time).  Note that the actual encoding
	and databases are not affected (just P2P protocol).
	The update is not complete yet, but should compile.

Tue Jan  8 20:07:20 MST 2008
	Added option for testing ("make check") to use weak(er)
	PRNG for key generation (thanks to Werner Koch for 
	pointing out how to do it).

Thu Dec 22 20:10:37 MST 2007
	Releasing GNUnet 0.7.3.

Sun Dec  9 14:34:32 MST 2007
	Implemented MySQL version of the dstore-module.  This
	means that sqLite is now again truly optional.

Sat Dec  8 15:15:53 MST 2007
	Integrated F2F topology into main topology module 
	(options are used to choose between F2F-only (old F2F), 
	 minimum number of friend connections (NEW) or entirely
	 arbitrary connection set (default)).

Thu Dec  6 20:51:17 MST 2007
	The SMTP transport is working again.

Thu Nov 22 20:49:16 MST 2007
	United all libgnunetutil_*.so libraries into one big
	libgnunetutil.so library (eliminates issues for binary

src/server/gnunetd.c
/*
     This file is part of GNUnet.
     (C) 2001, 2002, 2004, 2005, 2006 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file server/gnunetd.c
 * @brief Daemon that must run on every GNUnet peer.
 * @author Christian Grothoff
 * @author Larry Waldo
 * @author Tzvetan Horozov
 * @author Nils Durner
 */

#include "gnunet_util.h"
#include "gnunet_core.h"
#include "gnunet_directories.h"
#include "core.h"
#include "connection.h"
#include "tcpserver.h"
#include "handler.h"
#include "startup.h"
#include "version.h"

static struct GNUNET_GC_Configuration *cfg;
static struct GNUNET_GE_Context *ectx = NULL;

static struct GNUNET_CronManager *cron;
static struct GNUNET_LoadMonitor *mon;

#define PIDFILE_DATA "GNUNETD", "PIDFILE", GNUNET_DEFAULT_DAEMON_VAR_DIRECTORY "/gnunetd/pid"

static char *cfgFilename = GNUNET_DEFAULT_DAEMON_CONFIG_FILE;

static int no_daemonize_flag;

static int debug_flag;

static int loud_flag;

#ifndef WINDOWS
/**
 * Cron job that triggers re-reading of the configuration.
 */
static void
reread_config_helper (void *unused)
{
  GNUNET_GE_ASSERT (NULL, cfgFilename != NULL);
  if (0 != GNUNET_GC_parse_configuration (cfg, cfgFilename))
    GNUNET_GE_LOG (ectx,
		   GNUNET_GE_ERROR | GNUNET_GE_USER | GNUNET_GE_IMMEDIATE,
		   _("Failed to parse configuration file `%s'\n"),
		   cfgFilename);
}

/**
 * Signal handler for SIGHUP.
 * Re-reads the configuration file.
 */
static void
reread_config ()
{
  GNUNET_cron_add_job (cron, &reread_config_helper, 1 * GNUNET_CRON_SECONDS,
                       0, NULL);
}
#endif

/**
 * Park main thread until shutdown has been signaled.
 */
static void
waitForSignalHandler (struct GNUNET_GE_Context *ectx)
{
  GNUNET_GE_LOG (ectx,
                 GNUNET_GE_INFO | GNUNET_GE_USER | GNUNET_GE_REQUEST,
                 _("`%s' startup complete.\n"), "gnunetd");
  GNUNET_shutdown_wait_for ();
  GNUNET_GE_LOG (ectx,
                 GNUNET_GE_INFO | GNUNET_GE_USER | GNUNET_GE_REQUEST,
                 _("`%s' is shutting down.\n"), "gnunetd");
}

/**
 * The main method of gnunetd.
 */
int
gnunet_main ()
{
  struct GNUNET_SignalHandlerContext *shc_hup;
  int filedes[2];               /* pipe between client and parent */

  if ((GNUNET_NO == debug_flag) && (GNUNET_NO == no_daemonize_flag)
      && (GNUNET_OK != GNUNET_terminal_detach (ectx, cfg, filedes,
                                               PIDFILE_DATA)))
    return GNUNET_SYSERR;
  if ((GNUNET_NO != debug_flag) || (GNUNET_NO != no_daemonize_flag))
    GNUNET_pid_file_write (ectx, cfg, (unsigned int) getpid (), PIDFILE_DATA);
  if (NULL == (mon = GNUNET_network_monitor_create (ectx, cfg)))
    {
      if ((GNUNET_NO == debug_flag) && (GNUNET_NO == no_daemonize_flag))
        GNUNET_terminal_detach_complete (ectx, filedes, GNUNET_NO);
      else
        GNUNET_pid_file_delete (ectx, cfg, PIDFILE_DATA);
      return GNUNET_SYSERR;
    }
  cron = GNUNET_cron_create (ectx);
  GNUNET_GE_ASSERT (ectx, cron != NULL);
#ifndef WINDOWS
  shc_hup = GNUNET_signal_handler_install (SIGHUP, &reread_config);
#endif
  if (GNUNET_OK != GNUNET_CORE_init (ectx, cfg, cron, mon))
    {
      GNUNET_GE_LOG (ectx,
                     GNUNET_GE_FATAL | GNUNET_GE_USER | GNUNET_GE_IMMEDIATE,
                     _("Core initialization failed.\n"));

      GNUNET_cron_destroy (cron);
      GNUNET_network_monitor_destroy (mon);
#ifndef WINDOWS
      GNUNET_signal_handler_uninstall (SIGHUP, &reread_config, shc_hup);
#endif
      if ((GNUNET_NO == debug_flag) && (GNUNET_NO == no_daemonize_flag))
        GNUNET_terminal_detach_complete (ectx, filedes, GNUNET_NO);
      return GNUNET_SYSERR;
    }

  /* enforce filesystem limits */
  GNUNET_CORE_startup_cap_fs_quota_size (ectx, cfg);

  GNUNET_CORE_connection_init (ectx, cfg, mon, cron);
  GNUNET_CORE_load_application_modules ();
  if ((GNUNET_NO == debug_flag) && (GNUNET_NO == no_daemonize_flag))
    GNUNET_terminal_detach_complete (ectx, filedes, GNUNET_YES);
  GNUNET_cron_start (cron);
  GNUNET_CORE_p2p_enable_processing ();
  waitForSignalHandler (ectx);
  GNUNET_CORE_p2p_disable_processing ();
  GNUNET_cron_stop (cron);
  GNUNET_CORE_stop_cs_server ();
  GNUNET_CORE_unload_application_modules ();
  GNUNET_CORE_connection_done ();
  GNUNET_CORE_done ();
  GNUNET_network_monitor_destroy (mon);
#ifndef WINDOWS
  GNUNET_signal_handler_uninstall (SIGHUP, &reread_config, shc_hup);
#endif
  GNUNET_cron_destroy (cron);
  return GNUNET_OK;
}

#ifdef MINGW
/**
 * Main method of the windows service
 */
void WINAPI
ServiceMain (DWORD argc, LPSTR * argv)
{
  GNUNET_CORE_w32_service_main (gnunet_main);
}
#endif

/**
 * All gnunetd command line options
 */
static struct GNUNET_CommandLineOption gnunetdOptions[] = {
  GNUNET_COMMAND_LINE_OPTION_CFG_FILE (&cfgFilename),   /* -c */
  {'@', "win-service", NULL, "", 0,
   &GNUNET_getopt_configure_set_option, "GNUNETD:WINSERVICE"},
  {'n', "no-daemonize", NULL,
   gettext_noop ("do not daemonize (run gnunetd as a foreground process)"),
   0, &GNUNET_getopt_configure_set_one, &no_daemonize_flag},
  {'d', "debug", NULL,
   gettext_noop ("run in debug mode; gnunetd will "
                 "not daemonize and error messages will "
                 "be written to stderr instead of a logfile"),
   0, &GNUNET_getopt_configure_set_one, &debug_flag},
  GNUNET_COMMAND_LINE_OPTION_HELP (gettext_noop ("Starts the gnunetd daemon.")),        /* -h */
  GNUNET_COMMAND_LINE_OPTION_LOGGING,   /* -L */
  {'p', "padding-disable", "YES/NO",
   gettext_noop ("disable padding with random data (experimental)"), 0,
   &GNUNET_getopt_configure_set_option, "GNUNETD-EXPERIMENTAL:PADDING"},
  {'l', "loud", NULL,
   gettext_noop
   ("print all log messages to the console (only works together with -d)"),
   0, &GNUNET_getopt_configure_set_one, &loud_flag},
#ifndef MINGW
  {'u', "user", "USERNAME",
   gettext_noop ("specify username as which gnunetd should run"), 1,
   &GNUNET_getopt_configure_set_option, "GNUNETD:USERNAME"},
#endif
  GNUNET_COMMAND_LINE_OPTION_VERSION (PACKAGE_VERSION), /* -v */
  GNUNET_COMMAND_LINE_OPTION_END,
};

/**
 * Initialize util (parse command line, options) and
 * call the main routine.
 */
int
main (int argc, char *const *argv)
{
  int ret;

  if ((4 != sizeof (GNUNET_MessageHeader))
      || (600 != sizeof (GNUNET_MessageHello)))
    {
      fprintf (stderr,
               "Sorry, your C compiler did not properly align the C structs. Aborting.\n");
      return -1;
    }
  ret = GNUNET_init (argc,
                     argv,
                     "gnunetd [OPTIONS]",
                     &cfgFilename, gnunetdOptions, &ectx, &cfg);
  if (ret == -1)
    {
      GNUNET_fini (ectx, cfg);
      return 1;
    }
  GNUNET_pid_file_write (ectx, cfg, getpid (), PIDFILE_DATA);
  if (GNUNET_OK != GNUNET_CORE_startup_change_user (ectx, cfg))
    {
      GNUNET_pid_file_delete (ectx, cfg, PIDFILE_DATA);
      GNUNET_fini (ectx, cfg);
      return 1;
    }
  if (GNUNET_OK != GNUNET_CORE_startup_check_permissions (ectx, cfg))
    {
      GNUNET_pid_file_delete (ectx, cfg, PIDFILE_DATA);
      GNUNET_fini (ectx, cfg);
      return 1;
    }
  if (GNUNET_YES == debug_flag)
    {
      int dev;
      char *user_log_level;
      GNUNET_GE_KIND ull;

      GNUNET_GE_setDefaultContext (NULL);
      GNUNET_GE_free_context (ectx);
      GNUNET_GC_get_configuration_value_string (cfg,
                                                "LOGGING",
                                                "USER-LEVEL",
                                                "WARNING", &user_log_level);
      dev = GNUNET_GC_get_configuration_value_yesno (cfg,
                                                     "LOGGING", "DEVELOPER",
                                                     GNUNET_NO);
      ull = GNUNET_GE_getKIND (user_log_level);
      ull |= (ull - 1);         /* set bits for all lower log-levels */
      if (dev == GNUNET_YES)
        ull |= GNUNET_GE_DEVELOPER | GNUNET_GE_REQUEST;
      if (loud_flag == 1)
        ectx = GNUNET_GE_create_context_stderr (GNUNET_YES, GNUNET_GE_ALL);
      else
        ectx = GNUNET_GE_create_context_stderr (GNUNET_YES,
                                                GNUNET_GE_USER |
                                                GNUNET_GE_ADMIN | ull |
                                                GNUNET_GE_BULK |
                                                GNUNET_GE_IMMEDIATE);
      GNUNET_GE_setDefaultContext (ectx);
      GNUNET_free (user_log_level);
    }
  GNUNET_CORE_startup_set_fd_limit (ectx, cfg);
  if (GNUNET_OK != GNUNET_CORE_version_check_up_to_date (ectx, cfg))
    {
      GNUNET_GE_LOG (ectx,
                     GNUNET_GE_USER | GNUNET_GE_FATAL | GNUNET_GE_IMMEDIATE,
                     _
                     ("Configuration or GNUnet version changed.  You need to run `%s'!\n"),
                     "gnunet-update");
      GNUNET_pid_file_delete (ectx, cfg, PIDFILE_DATA);
      GNUNET_fini (ectx, cfg);
      return 1;
    }

#ifdef MINGW
  if (GNUNET_GC_get_configuration_value_yesno (cfg,
                                               "GNUNETD", "WINSERVICE",
                                               GNUNET_NO) == GNUNET_YES)
    {
      SERVICE_TABLE_ENTRY DispatchTable[] = { {"GNUnet", ServiceMain}
      , {NULL, NULL}
      };
      ret = (GNStartServiceCtrlDispatcher (DispatchTable) != 0);
    }
  else
#endif
    ret = gnunet_main ();
  GNUNET_pid_file_delete (ectx, cfg, PIDFILE_DATA);
  GNUNET_fini (ectx, cfg);
  if (ret != GNUNET_OK)
    return 1;
  return 0;
}

/* You have reached the end of GNUnet. You can shutdown your
   computer and get a life now. */
